lisa/tools/fips.py

```diff
@@ -0,0 +1,379 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+import re
+from abc import ABC, abstractmethod
+from pathlib import PurePosixPath
+from typing import TYPE_CHECKING, Any
+
+from assertpy import assert_that
+
+from lisa.executable import Tool
+from lisa.operating_system import CBLMariner, OperatingSystem
+from lisa.tools import Blkid, Cat, Sed
+from lisa.util import UnsupportedDistroException, get_matched_str
+
+if TYPE_CHECKING:
+    from lisa.operating_system import Posix
+    from lisa.node import Node
+
+
+def assert_package_existance(
+    os: OperatingSystem, package: str, should_exist: bool
+) -> None:
+    """
+    Assert that a package/tool is installed on the node.
+    """
+    assert_that(os.package_exists(package)).described_as(
+        f"Package {package} is installed."
+    ).is_equal_to(should_exist)
+
+
+class Grub(Tool, ABC):
+    @classmethod
+    def create(cls, node: "Node", *args: Any, **kwargs: Any) -> Tool:
+        """
+        Factory method to create an instance of the Grub tool.
+        """
+        if isinstance(node.os, CBLMariner):
+            if node.os.information.release == "2.0":
+                return GrubAzl2(node, args, kwargs)
+            if node.os.information.release == "3.0":
+                return GrubAzl3(node, args, kwargs)
+
+        raise UnsupportedDistroException(
+            os=node.os, message="Grub tool only supported on CBLMariner 2.0 and 3.0."
+        )
+
+    def __init__(
+        self, command: str, package: str, node: "Node", *args: Any, **kwargs: Any
+    ) -> None:
+        super().__init__(node, *args, **kwargs)
+        self._command = command
+        self._package = package
+
+    @property
+    def command(self) -> str:
+        return self._command
+
+    @property
+    def can_install(self) -> bool:
+        return True
+
+    @abstractmethod
+    def set_fips_mode(self, fips_mode: bool) -> None:
+        """
+        Set the FIPS mode to the specified value.
+        """
+
+    def set_boot_uuid(self, uuid: str, same_as_root: bool) -> None:
+        """
+        Set the boot UUID to the specified value.
+        """
+        if same_as_root:
+            self.remove_kernel_cmdline_arg(r"boot")
+        else:
+            self.set_kernel_cmdline_arg(f"boot=UUID={uuid}")
+
+    @abstractmethod
+    def unset_boot_uuid(self, same_as_root: bool) -> None:
+        """
+        Unset the boot UUID.
+        """
+
```

Comment:
We don't use `abstractmethod` in LISA, because sometimes it has some problems on reflection. So, use normal methods and raise NotImplementedError inside.